### PR TITLE
Roll call from activities [Event Attendance Part 3]

### PR DESCRIPTION
**Pull request by sleidig**
see issue: #495

### Visible/Frontend Changes
- [x] redesigned Roll Call feature, displaying available recurring activities and one-time events for selection

### Architectural/Backend Changes
- [x] `Note` entities can link to 
- [x] adapted root layout heights to allow components to base their size on relative parents

### TODOs
- [x] save `Note.relatesTo` reference to DB (but only the id, not the whole related entity)
- [x] implement "My event is not listed", showing popup to create a new Note (i.e. one-time event)
- [x] do not show events (notes) which are not "meeting" notes
- [x] filter activities in roll call setup: only display those assigned to current user, show other activities if the user clicks "show more"

> Further steps will follow in additional PRs, all collected in an "Epic" PR before merging collectively into master.
> Review code only after PR #560 is merged into epic/event-attendance (changes build upon each other)

### CODE REVIEW

**Comment by TheSlimvReal** [comment]
Link to the deployment: https://event-attendance-3-test-deploy.herokuapp.com/

File: src/app/child-dev-project/children/children.module.ts [hunk 1]
```diff
@@ -125,6 +120,7 @@ import { EntityListModule } from "../../core/entity-components/entity-list/entit
     ViewModule,
     EntitySubrecordModule,
     EntityListModule,
+    EntityDetailsModule,
```
Comment (TheSlimvReal) [inline src/app/child-dev-project/children/children.module.ts:123]: Why is this needed?

File: src/app/child-dev-project/notes/model/note.ts [hunk 2]
```diff
@@ -15,26 +15,44 @@
...
-  @DatabaseField({ dataType: "interaction-type" }) category: InteractionType = {
-    name: "",
-  };
+  @DatabaseField({ dataType: "interaction-type" }) category: InteractionType =
+    InteractionType.NONE;
+
+  /**
+   * reference to a different entity (e.g. a recurring activity) this note is related to
+   */
+  relatesTo: Entity; // TODO: DatabaseField !! using an entity loader index  id <-> entity instance
```
Comment (TheSlimvReal) [inline src/app/child-dev-project/notes/model/note.ts:54]: Is there a reason why we cant just put the ID of the activity here? I only see it used when looking for events that belong to a certain activity.

File: src/app/core/ui/ui/ui.component.scss [hunk 3]
```diff
@@ -54,12 +55,13 @@ app-ui {
 
 .sidenav-content {
   padding: 24px;
-  /* overflow: scroll; */
+  height: calc(100vh - 114px);
```
Comment (TheSlimvReal) [inline src/app/core/ui/ui/ui.component.scss:58]: A short comment what this does would be helpful

**Review by TheSlimvReal — CHANGES_REQUESTED** [review]
Very solid work. Feels already very easy to use. I also like how easy some components got. I think we can even reduce the complexity a little bit more.

File: src/app/child-dev-project/attendance/activity-setup/activity-setup.component.html [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    <div fxLayout="row" fxLayoutGap="16px">
+      <button mat-raised-button color="primary" class="confirm-button"
+              [disabled]="!selectedEvent"
+              (click)="eventSelected.emit(selectedEvent)">
+        Start Roll Call
+      </button>
+
+      <app-activity-card *ngIf="selectedEvent" [event]="selectedEvent"
+                         class="mat-elevation-z0 selected-card" fxFlex>
+      </app-activity-card>
```
[the remark this answers is not in the document]
Comment (sleidig) [reply src/app/child-dev-project/attendance/activity-setup/activity-setup.component.html:40]: The challenge is that the cards can have different height at the moment, depending on line breaks and what information is available. I kept it like this because I felt it more important to keep the bottom panel at a fixed size while not "wasting" too much valuable screen space. Not ideal, I agree, but the best thing I could come up with till now.

File: src/app/child-dev-project/attendance/activity-setup/activity-setup.component.ts [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    await this.loadActivities();
+    this.sortEvents();
+  }
+
+  private async loadActivities() {
+    this.allActivities = await this.entityService.loadType<RecurringActivity>(
+      RecurringActivity
+    );
+    // TODO: smart filters based on assignedTo, date patterns, etc.
+    this.visibleActivities = this.allActivities;
```
[the remark this answers is not in the document]
Comment (sleidig) [reply src/app/child-dev-project/attendance/activity-setup/activity-setup.component.ts:45]: now implemented using `RecurringActivity.assignedTo`. We can extend this with more complex logic later on

File: src/app/child-dev-project/attendance/add-day-attendance/add-day-attendance.component.ts [hunk 6]
```diff
@@ -1,33 +1,50 @@
...
+
+  constructor(
+    private entityService: EntityMapperService,
+    private childrenService: ChildrenService
+  ) {}
+
+  ngOnInit(): void {
+    this.childrenService
+      .getChildren()
+      .subscribe((children) => (this.allChildren = children));
```
[the remark this answers is not in the document]
Comment (sleidig) [reply src/app/child-dev-project/attendance/add-day-attendance/add-day-attendance.component.ts:33]: with EntityMapperService the Child objects do not have their photo, which is loaded and injected by ChildrenService during load. (something we need to solve with a schema datatype somehow ...)

File: src/app/child-dev-project/children/child-select/child-select.component.html [hunk 7]
```diff
@@ -1,10 +1,12 @@
...
-           [(ngModel)]='searchText'>
+    <input #inputField matInput placeholder='Children' (input)='search()'
+           [matAutocomplete]='autoSuggestions'
+           [(ngModel)]='searchText'
+           [disabled]="disabled">
     <span *ngFor='let child of selectedChildren' class="selected-entity-block">
         <span class='fa fa-minus-square-o' (click)='unselectChild(child)'></span>
-        <app-child-block [entity]='child' [linkDisabled]='true' [tooltipDisabled]="true"></app-child-block>
+        <app-child-block [entity]='child' [linkDisabled]='true' [tooltipDisabled]="!disabled"></app-child-block>
         &nbsp;
```
[the remark this answers is not in the document]
Comment (sleidig) [reply src/app/child-dev-project/children/child-select/child-select.component.html]: fair points, would you open new GitHub issues for that? I am just trying to slightly adapt the component here to make it usable for the current use case.

File: src/app/child-dev-project/notes/model/note.ts (hunk 2 above)
Comment (sleidig) [reply src/app/child-dev-project/notes/model/note.ts:54]: I am using the referenced activity object now to check whether the event is assigned to the current user. Want to avoid having to load these from the db separately again.

Maybe the pain is big enough for me to work on #228  in order to resolve this

File: src/app/child-dev-project/attendance/add-day-attendance/add-day-attendance.component.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (sleidig) [reply src/app/child-dev-project/attendance/add-day-attendance/add-day-attendance.component.ts:33]: actually it's not the photo, but the current school that is loaded additionally. Anyway, ... I created an issue to think about making this obsolete: #581

File: src/app/child-dev-project/children/child-select/child-select.component.html (hunk 7 above)
[the remark this answers is not in the document]
Comment (sleidig) [reply src/app/child-dev-project/children/child-select/child-select.component.html:10]: see #577

File: src/app/child-dev-project/notes/model/note.ts (hunk 2 above)
Comment (sleidig) [reply src/app/child-dev-project/notes/model/note.ts:54]: worked around it and found a pretty good solution, I hope.

File: src/app/child-dev-project/attendance/activity-setup/activity-setup.component.ts [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+      return score;
+    };
+
+    this.existingEvents = this.existingEvents.sort(
+      (a, b) => calculateEventPriority(b) - calculateEventPriority(a)
+    );
+  }
+
+  createOneTimeEvent() {
+    const newNote = new Note(Date.now().toString());
```
Comment (TheSlimvReal) [inline src/app/child-dev-project/attendance/activity-setup/activity-setup.component.ts:122]: Is there a rule to create notes like this? With the date as the id?

File: src/app/child-dev-project/attendance/add-day-attendance/add-day-attendance.component.ts [hunk 9]
```diff
@@ -13,48 +12,20 @@ export class AddDayAttendanceComponent {
...
   event: Note;
 
-  stages = ["Setup Roll Call", "Select Student Group", "Roll Call"];
+  stages = ["Select Event", "Record Attendance"];
 
-  constructor(private entityService: EntityMapperService) {}
+  constructor(private entityMapper: EntityMapperService) {}
 
-  finishBasicInformationStage() {
+  async finishBasicInformationStage(event: Note) {
```
Comment (TheSlimvReal) [inline src/app/child-dev-project/attendance/add-day-attendance/add-day-attendance.component.ts:22]: why async?

File: src/app/child-dev-project/attendance/add-day-attendance/roll-call/roll-call.component.html [hunk 10]
```diff
@@ -14,8 +14,15 @@
...
 <div *ngIf='isFinished()' class='roll-call-complete' [@completeRollCall]>
-  <span class='fa fa-check-circle-o roll-call-complete-icon'></span><br>
-  Roll call completed.
+  <span class='fa fa-check-circle-o roll-call-complete-icon'></span>
+  <div>
+    Roll call completed.
+  </div>
+  <div>
+    <button mat-raised-button color="primary" class="roll-call-complete-save nav-button" (click)="finish()" fxFlex>
+      <b>Save</b>
```
Comment (TheSlimvReal) [inline src/app/child-dev-project/attendance/add-day-attendance/roll-call/roll-call.component.html:23]: Is there a reason why it is not immediately saved, once the roll call is finished?

File: src/app/child-dev-project/attendance/attendance.service.ts [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+import { Note } from "../notes/model/note";
+
+@Injectable({
+  providedIn: "root",
+})
+export class AttendanceService {
+  constructor(private entityMapper: EntityMapperService) {}
+
+  async getEventsOnDate(date: Date) {
+    const events = await this.entityMapper.loadType<Note>(Note);
```
Comment (TheSlimvReal) [inline src/app/child-dev-project/attendance/attendance.service.ts:13]: This could definitely be speed up by a index which is sorted by the event date.

**Review by TheSlimvReal — APPROVED** [review]
Looks very solid now, I just added some comments for clarification

File: src/app/child-dev-project/attendance/activity-setup/activity-setup.component.ts (hunk 8 above)
Comment (sleidig) [reply src/app/child-dev-project/attendance/activity-setup/activity-setup.component.ts:122]: `Note.create(new Date());` seems more elegant. will change it

File: src/app/child-dev-project/attendance/add-day-attendance/roll-call/roll-call.component.html (hunk 10 above)
Comment (sleidig) [reply src/app/child-dev-project/attendance/add-day-attendance/roll-call/roll-call.component.html:23]: That's a good idea!

File: src/app/child-dev-project/attendance/attendance.service.ts (hunk 11 above)
Comment (sleidig) [reply src/app/child-dev-project/attendance/attendance.service.ts:13]: that's on the todo list but maybe deserves its on PR
